base/server/python/pki/server/deployment/scriptlets/selinux_setup.py

```diff
@@ -154,67 +154,57 @@ def destroy(self, deployer):
-                            "deleting selinux fcontext \"%s\"",
-                            deployer.mdict['pki_server_database_path'] + self.suffix)
-                        fcon.delete(
-                            deployer.mdict['pki_server_database_path'] +
-                            self.suffix, "")
-
-                        port_records = seobject.portRecords(trans)
-                        for port in ports:
-                            logger.info("deleting selinux port %s", port)
-                            port_records.delete(port, "tcp")
-
-                    trans.finish()
-                break
-            except ValueError as e:
-                error_message = str(e)
-                logger.error(error_message)
-                if error_message.strip() == \
-                        "Could not start semanage transaction":
-                    counter += 1
-                    if counter >= max_tries:
-                        raise
-                    time.sleep(5)
-                    logger.debug("Retrying to remove selinux context ...")
-                else:
-                    raise
+                    trans = seobject.semanageRecords("targeted")
+                    try:
+                        trans.start()
+                        # If a port context is specified
+                        if context == 'ports':
+                            port_records = seobject.portRecords(trans)
+                            for port in ports:
+                                logger.info("deleting selinux port %s", port)
+                                port_records.delete(port, "tcp")
+
+                        # Else it's a file context
+                        else:
+                            fcon = seobject.fcontextRecords(trans)
+
+                            logger.info(
+                                "deleting selinux fcontext \"%s\"",
+                                deployer.mdict[context] + self.suffix)
+                            fcon.delete(
+                                deployer.mdict[context] + self.suffix, "")
+                        break
+                    except ValueError as e:
+                        error_message = str(e)
+                        logger.error(error_message)
+                        if error_message.strip() == \
+                                "Could not start semanage transaction":
+                            if counter >= max_tries:
+                                raise
+                            time.sleep(5)
+                            logger.debug("Retrying to remove selinux context ...")
+                        else:
+                            # If it is a forced destroy, there won't be any se contexts to destroy
+                            if deployer.mdict['pki_force_destroy']:
+                                break
+                            else:
+                                raise
+                    finally:
+                        trans.finish()
```

Comment:
I'm not an expert in SELinux, but I suppose the actual operations only begin once the transaction is committed by calling trans.finish(), so if we want to catch any failure the trans.finish() should remain inside the try-catch block.

Also, in the original code all operations are executed in the same transaction. If it fails, the entire transaction is cancelled then retried again (it's all or nothing). In the new code each SELinux operation is executed as a separate transaction which may result in inconsistent state.

Reply:
> actual operations only begin once the transaction is committed by calling trans.finish()

I can't find docs for the SELinux Python API. So, I looked at the source code. As you can see in [1] and [2], there is a `commit()` at the end of `delete()` which deletes the context/entries from the SElinux db. Moreover `trans.finish()` [3] marks a SElinux transaction to be completed and commits (by calling the previous method) the changes again. I understood this more like a real database transaction.

Moreover, if there is an exception, the transaction never gets marked as finish which leads to "Too many files open" error -- an issue I saw in my dev machine.

> Also, in the original code all operations are executed in the same transaction

That is correct. I was worried about how it did that and so refactored this. Consider the following scenairos:

| Scenario | Old Code | New Code |
|-------------|-------------|---------------|
| Normal destroy | Removes all port and file context | <no change>|
| **destroy with "Could not start semanage transaction"**| **10 Retries** | **10retries for each context** |
| Destroying broken installation | Tries to remove 1 by 1 context. If not found, the other contexts are skipped | Ensure it removes all contexts |

IIUC, your concern was about scenario 2. Suppose there is a different error/exception during execution of `.delete()`, the old code will (a) not end a transaction (b) not remove the other contexts. While, the new code will (a) cleanly end the transaction (b) will make sure to remove all contexts.

[1] https://github.com/SELinuxProject/selinux/blob/master/python/semanage/seobject.py#L1251
[2] https://github.com/SELinuxProject/selinux/blob/master/python/semanage/seobject.py#L2599
[3] https://github.com/SELinuxProject/selinux/blob/master/python/semanage/seobject.py#L336

We can disucss over IRC if you need any more clarification :)

Reply:
IIUC the semanage code supports both transactional and non-transactional mode. In non-transactional mode, each operation will create a transaction for that operation only using begin() and commit() as in [1]. In transactional mode the transaction is started with start(). The begin() and commit() will simply return, doing nothing. Then in finish() the actual commit is executed.

My concern is actually scenario #4, pkidestroy gets interrupted (due to internal bug or external interrupt) so some contexts are removed but some others are not since they are in separate transactions. In the old code this situation will not happen since with a single transaction it's either all or nothing (i.e. scenario #3 never happens).

If there's a "Too many files open" error I think it should be addressed separately. It's outside the scope of PKI tools.

Reply:
As discussed on IRC, for cleanup purposes the non-transactional mode would actually be better since that way pkidestroy can be executed multiple times if necessary, each time removing as many contexts as possible, and eventually finishes with rc=0 when none of the contexts exist anymore.

Reply:
Yuck, can we please make it do the loop internally instead of aborting?

I'm reminded of the nonsense I have to do for podman because it does the same thing:

    ret=1 ; while (( ret != 0 )); podman rmi --force --all ; ret=$? ; done

For nested build; they try and remove it and if it fails because of a dependency, they skip it. Eventually you call rmi enough times that everything gets removed cleanly.

Reply:
Just to clarify, the behavior I'm asking for is:

Try to remove/cleanup everything. If anything failed, show what failed at the end. If it is an exceptional error that we can't just defer handling, then exit (e.g., SEGV, ctrl+c, IO error, etc). i.e., rerunning might be helpful, but might not be. 

I understood the behavior that was asked for is:

For each item, remove, if error, exit immediately. It is up to the user to rerun until they get a clean system.

Reply:
So, I'm trying to understand the motivation here.

Here is how I understand how I wrote the original code.  The code would start a transaction in which all the required contexts were removed together.  The retry mechanism was there because sometimes the selinux engine was busy and so the transaction could not be started.  So, if the transaction could not be started, then we retried 10 times (with a sleep in between).  If the transaction started and it failed for some other reason, then an exception was raised.

In this new code, we have the same retry mechanism for transactions not starting, but this time we do each context removal as a separate transaction.  This leads to the possibility of having some contexts being removed and others not, which could be confusing.

As far as I know, there is no case of some contexts being removed and others not.  Is there?

Reply:
With the current code, if the contexts were complete initially the transaction should succeed, assuming there's no selinux engine issue. However, if the contexts were already incomplete or messed up, the transaction could fail and nothing gets removed. I don't think we should assume that the existing contexts are complete due to possible manual changes or other bugs. So if possible I'd suggest checking each context first whether it exists before trying to remove it, regardless with transaction or not.

Reply:
oh there is 1 more primary reason for this reformat. With the introduction of `--force`, pkidestroy tries to execute all the scriptlets. As a result, it was raising exception as soon as it tried to remove the first non-existent context.

For the transactional vs non-transactional model, both worked as expected. So, I don't have any objections to remove the `trans.{start, finish}()` function calls.

Reply:
Agree on the check t see if the context exists.

Personally, I would leave this as a single transaction because it makes things much clearer in general when things don't go wrong.  And we can address the issues raised above by adding a check if the context exists.

Another thing to keep in mind if you separate the transactions. If the selinux engine requires at least one retry, then you're going to add wait time for every context removal - adding to the overall operation time.  We may not care too much when destroying a single instance - but IPA (which has to clean up a bunch of things) might...